library/core/src/ffi/mod.rs

```diff
@@ -92,56 +92,46 @@ pub type c_ssize_t = isize;
 mod c_char_definition {
     cfg_if! {
         // These are the targets on which c_char is unsigned.
-        if #[cfg(any(
-            all(
-                target_os = "linux",
-                any(
-                    target_arch = "aarch64",
-                    target_arch = "arm",
-                    target_arch = "hexagon",
-                    target_arch = "powerpc",
-                    target_arch = "powerpc64",
-                    target_arch = "s390x",
-                    target_arch = "riscv64",
-                    target_arch = "riscv32",
-                    target_arch = "csky"
-                )
+        // See isSignedCharDefault() in clang/lib/Driver/ToolChains/Clang.cpp
+        // PowerPC uses unsigned char for all targets except Darwin.
+        // Arm/AArch64 uses unsigned char except for Darwin and Windows targets
+        if #[cfg(all(
+            any(
+                target_os = "ios",
+                target_os = "macos",
+                target_os = "tvos",
+                target_os = "visionos",
+                target_os = "watchos",
             ),
-            all(target_os = "android", any(target_arch = "aarch64", target_arch = "arm")),
-            all(target_os = "l4re", target_arch = "x86_64"),
-            all(
-                any(target_os = "freebsd", target_os = "openbsd", target_os = "rtems"),
-                any(
-                    target_arch = "aarch64",
-                    target_arch = "arm",
-                    target_arch = "powerpc",
-                    target_arch = "powerpc64",
-                    target_arch = "riscv64"
-                )
+            any(
+                target_arch = "aarch64",
+                target_arch = "arm",
+                target_arch = "powerpc",
+                target_arch = "powerpc64",
             ),
-            all(
-                target_os = "netbsd",
-                any(
-                    target_arch = "aarch64",
-                    target_arch = "arm",
-                    target_arch = "powerpc",
-                    target_arch = "riscv64"
-                )
-            ),
-            all(
-                target_os = "vxworks",
-                any(
-                    target_arch = "aarch64",
-                    target_arch = "arm",
-                    target_arch = "powerpc64",
-                    target_arch = "powerpc"
-                )
-            ),
-            all(
-                target_os = "fuchsia",
-                any(target_arch = "aarch64", target_arch = "riscv64")
+        ))] {
+            pub type c_char = i8;
+        } else if #[cfg(all(
```

Comment:
We should probably only have one `if` and one `else`, one for `u8`, one for `i8`, having two `i8` branches should at worst be done by `any(`ing these two expressions together. We definitely shouldn't have 4 branches, 3 i8 and 1 u8.

Reply:
I did it this way to avoid duplicating the 5 darwin OS combinations but as you say it might be cleaner to do it that way. Let me try that

Reply:
I simplified this by relying on Apple operating systems other than macos only supporting Arm, so it's not too bad.

Reply:
Thanks this is so much simpler!